File: rotate_array.java
// Time: O(n) | Space: O(1)
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        if (n == 0) return;
        k = ((k % n) + n) % n;
        reverse(nums, 0, n - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, n - 1);
    }

    private static void reverse(int[] a, int l, int r) {
        while (l < r) {
            int tmp = a[l];
            a[l] = a[r];
            a[r] = tmp;
            l++; r--;
        }
    }
}

File: rotate_array.py
# Time: O(n) | Space: O(1)
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        if n == 0: return
        k %= n
        self.rvrs(nums, 0, n - 1)
        self.rvrs(nums, 0, k - 1)
        self.rvrs(nums, k, n - 1)

    def rvrs(self, a: List[int], l: int, r: int) -> None:
        while(l < r):
            a[l], a[r] = a[r], a[l]
            l+=1
            r-=1
